tls/src/server.rs

```diff
@@ -249,176 +250,193 @@ impl Server {
     }
 
     pub async fn accept(
-        &mut self,
-        corpus: &[Utf8PathBuf],
-    ) -> Result<(Stream<TcpStream>, core::net::SocketAddr), Error> {
-        // load corims into a set of ReferenceMeasurements
-        let mut corims = Vec::new();
-        for c in corpus {
-            corims.push(Corim::from_file(c)?);
-        }
-        let corpus = ReferenceMeasurements::try_from(corims.as_slice())?;
-
-        let (stream, addr) = self.tcp_listener.accept().await?;
-        let mut stream = self.tls_acceptor.clone().accept(stream).await?;
-
-        // get PlatformId from server TLS / Trust Quorum cert chain
-        let (_, conn) = stream.get_ref();
-        let tq_platform_id = if let Some(tls_certs) = conn.peer_certificates() {
-            let mut pki_path = Vec::new();
-            for der in tls_certs.iter() {
-                pki_path.push(Certificate::from_der(der).map_err(|_| {
-                    rustls::Error::InvalidCertificate(
-                        rustls::CertificateError::BadEncoding,
-                    )
-                })?)
+        &self,
+        corpus: Vec<Utf8PathBuf>,
+    ) -> impl Future<
+        Output = Result<(Stream<TcpStream>, core::net::SocketAddr), Error>,
+    > {
```

Comment:
This is kind of a strange API - if we're already async, why not just go ahead and do the remaining asynchronous work and let the caller `spawn()` the whole thing if they want to?

I assume this didn't work before because of `&mut self`, but if we're able to change it to `&self` - could we not return the nested future and still be spawnable?

Reply:
I agree this is a strange API, which is primarily why I tagged you on this review :) 

The problem is that we don't know when to spawn that task. The TCP accept is what indicates there is a connection and lets us know that we can proceed with the spawn to perform the handshake. This is why `tokio_rustls` takes the TCP stream itself as a parameter and then performs the TLS handshake. It separates the two by having the caller have to manage the TCP socket separately, whereas this code wraps the TCP accept by returning it's own future. 

We could have the caller spawn the whole things, but then we'd have to spawn some number of accept tasks and each time one went away, spawn another. This seems trickier and less friendly to the user.

Reply:
Good idea. Either of those would be less weird.

I originally wrapped the TCP accept so that the tcp socket couldn't be used on it's own before the sprockets handshake was complete. This was mainly to prevent mistakes where somehow we ended up using plain sockets other than attested TLS channels.  But thinking about that some more, I realize that's kind of silly. The connection side will always be sprockets only and so the handshake will bomb out. And plus, this is only used in 1 place. I say this, because I think the `tokio_rustls` version is a little clearer.